Flow - RDD to DF

rabarona opened the pull request:
Implements DataFrames for netflow related classes:

Moving word creation to FlowWordCreation.
Implementing FlowSchema.
Flow Feedback refactoring to return DataFrame instead of RDD.
Updated FlowSchema with intermediate columns.
Updated FlowWordCreationTest.

Commit:
Changed DNS Post and Pre LDA to use DF instead of RDDs.

Commit:
Merge branch 'dev' of https://github.com/Open-Network-Insight/oni-ml into 1.1-ml_RDD_to_DF

Commit:
Word creation taken out of DNSPre and Post

Commit:
Merge branch 'dev' of https://github.com/Open-Network-Insight/oni-ml into 1.1-ml_RDD_to_DF

Commit:
DataFrame implementation testing.

Commit:
Implemented DNSSchema

Commit:
Added missing column after DNSSchema implementation. Fixed select after word creation.

Commit:
Fixed scala.MatchError.

Commit:
Merge branch 'dev' of https://github.com/Open-Network-Insight/oni-ml into 1.1-ml_RDD_to_DF

# Conflicts:
#	src/main/scala/org/opennetworkinsight/dns/DNSPostLDA.scala
Conflict fixed git mergetool and then manually.

Commit:
Manual merge after PR #57.

Commit:
Fixed typo in line 146.

Commit:
Merge branch 'dev' of https://github.com/Open-Network-Insight/oni-ml into 1.1-ml_Flow_RDD_to_DF

Commit:
Updated DNSPostLDA score function to broadcast variables until they are needed.
FlowPre/PostLDA refactoring. Moving out word creation, implementing FlowSchema, changing RDD to DataFrames.
Flow Feedback refactoring to return DataFrame instead of RDD.
Updated FlowSchema with intermediate columns.
Updated FlowWordCreationTest.

daortizh commented on src/main/scala/org/opennetworkinsight/dns/DNSWordCreation.scala:
```diff
@@ -3,17 +3,32 @@ package org.opennetworkinsight.dns
 
   object DNSWordCreation {
 
+    val countryCodes = Set("ac", "ad", "ae", "af", "ag", "ai", "al", "am", "an", "ao", "aq", "ar", "as", "at", "au",
```
Is there any way to avoid having this set of country codes hardcoded? And if there is... is there any benefit?

rabarona replied:
I guess we can read from a file instead of this long variable. 
Will keep in mind for the future.

daortizh replied:
:+1:

NathanSegerlind replied:
https://jira01.devtools.intel.com/browse/DB-521

hahaahaha

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSPostLDA.scala:
```diff
@@ -24,15 +23,10 @@ object DNSPostLDA {
-
     val totalDataDF = {
-      sqlContext.parquetFile(inputPath.split(",")(0))
+      sqlContext.read.parquet(inputPath.split(",")(0))
```
thanks for updating the call to 1.6

rabarona replied:
No problem. I'm utilizing IntelliJ VersionControl feature. From VersionControl view I get a confirmation window before committing and pushing changes where it tells me all the warnings in the code, the parquetFile appeared as "will be deprecated in 2.0", that's how I did this change.

daortizh commented:
LGTM

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSPostLDA.scala:
```diff
@@ -24,15 +23,10 @@ object DNSPostLDA {
-
     val totalDataDF = {
-      sqlContext.parquetFile(inputPath.split(",")(0))
+      sqlContext.read.parquet(inputPath.split(",")(0))
```
I am having serious doubts about the role of CSVs and TSVs in the whole design... we're seeing fields that contain commas and that can mess things up.... 

it seems like parquet is handling them okay so far but I'd like to be really sure that there are no problems with  this going forward

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/dns/DNSPostLDA.scala:
```diff
@@ -48,7 +42,7 @@ object DNSPostLDA {
 
     logger.info("Computing conditional probability")
```
I'm gonna add a Jira ticket to "fix the logging functinoality in the ML pipeline"

https://jira01.devtools.intel.com/browse/DB-520

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/netflow/FlowPostLDA.scala:
```diff
@@ -19,112 +17,125 @@ object FlowPostLDA {
-          "rip")
-      flowDataFrame.map(_.mkString(","))
+        .select(Schema.TimeReceived,
+          Schema.Year,
```
what do you think of doing an import FlowSchema._  and then using Month, Duration, SourcePort etc as constants throughout the file?  

since the only schema names being used are the Flow ones, there's no confusion and it cuts back on a lot of redundant notation

rabarona replied:
That sounds good to me. Does your idea includes DNS, Proxy as well? or just flow because it's a long list?

NathanSegerlind replied:
I've tried it with proxy feedback and it looks really good, more concise and less clutter

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/netflow/FlowPreLDA.scala:
```diff
@@ -23,186 +21,126 @@ object FlowPreLDA {
-          "rip")
-      flowDataFrame.map(_.mkString(","))
-    }
+    val falsePositives: DataFrame = if (scoredFileExists) {
```
the more I look at this, the more I want to separate out the feedback into a separate object to hide its details from the main training routines of prelda...  most of the lines here are simply accounting for the format of the feedback file and it obscures the logic....

I will try it that way for the proxy feedback and we can see if it's worth doing the change for the other pipelines

rabarona replied:
It's worth doing that change, it's something I was tempting to do too but because of the other priorities I didn't. 

Will work on it next sprint for DNS and Flow.

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/netflow/FlowPreLDA.scala:
```diff
@@ -23,186 +21,126 @@ object FlowPreLDA {
 
-    val data_with_words = binned_data.map(row => FlowWordCreation.adjustPort(row))
+    val src_word_counts = dataWithWordDF.select(Schema.SourceIP, Schema.SourceWord)
+      .map({
```
the formatting of this map is confusing and it actually all fits easily into 120 character line

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/netflow/FlowPreLDA.scala:
```diff
@@ -23,186 +21,126 @@ object FlowPreLDA {
 
-    val src_word_counts = data_with_words.map(row => ((row(indexOf.SOURCEIP),  row(indexOf.SOURCEWORD)), 1)).reduceByKey(_ + _)
-    val dest_word_counts = data_with_words.map(row => ((row(indexOf.DESTIP),  row(indexOf.DESTWORD)), 1)).reduceByKey(_ + _)
+    val dest_word_counts = dataWithWordDF.select(Schema.DestinationIP, Schema.DestinationWord)
```
map statement could fit on one line

NathanSegerlind commented on src/main/scala/org/opennetworkinsight/netflow/FlowSchema.scala:
```diff
@@ -13,7 +13,7 @@ object FlowSchema {
   val Second = "trsec"
   val Duration = "tdur"
-  val SourceIP = "sip "
+  val SourceIP = "sip"
```
how did this work before???

doh... dev branch uses the string literals not the Schema object

rabarona replied:
you mean Flow worked with the changes you reverted later?

NathanSegerlind replied:
how it worked before: dev branch didn't use the Schema object

rabarona replied:
that's right.

NathanSegerlind commented on src/test/scala/org/opennetworkinsight/FlowWordCreationTest.scala:
```diff
@@ -12,63 +11,63 @@ class FlowWordCreationTest extends FlatSpec with Matchers {
-    result(FlowColumnIndex.IPKTYBIN) shouldBe "4"
-    result(FlowColumnIndex.TIMEBIN) shouldBe "8"
-  }
+  //  "binIbytIpktTime" should "count 5 ibyt_bin for ibyt 208, 3 ipkt_bin for ipkt 4 and 0 time_bin for time " +
```
we shouldn't be reducing test coverage... I guess something changed that made the test fail... can't we just update the test? or did the change introduce a mysterious change in results?

rabarona replied:
I should have removed this code as binIbytIpktTime doesn't exists anymore. 
We need to create more UnitTest for the new flowWordCreation function.

NathanSegerlind commented:
looks pretty good, ricardo

my comments are mostly cosmetic, although I am concerned about the commented out unit test....

if the test is now completely invalid, it should be deleted
if it's still valid, it should pass

Commit:
Removed out of data unit testing
Importing DNS/FlowSchema._ and use column names as constants.
Code style improvements.

rabarona commented:
In commit #1737753 I meant "Out of date". Sorry for the typo.

NathanSegerlind commented:
looking good
